// convex/dist/esm/cli/convexImport.js
"use strict";
import chalk from "chalk";
import {
  ensureHasConvexDependency,
  formatSize,
  waitUntilCalled,
  deploymentFetch,
  logAndHandleFetchError
} from "./lib/utils/utils.js";
import {
  logFailure,
  oneoffContext,
  showSpinner,
  logFinishedStep,
  logWarning,
  logMessage,
  stopSpinner,
  changeSpinner
} from "../bundler/context.js";
import {
  fetchDeploymentCredentialsProvisionProd,
  deploymentSelectionFromOptions
} from "./lib/api.js";
import path from "path";
import { subscribe } from "./lib/run.js";
import { Command, Option } from "@commander-js/extra-typings";
import { actionDescription } from "./lib/command.js";
import { ConvexHttpClient } from "../browser/http_client.js";
import { makeFunctionReference } from "../server/index.js";
import { deploymentDashboardUrlPage } from "./dashboard.js";
import { promptYesNo } from "./lib/utils/prompts.js";
const CHUNK_SIZE = 5 * 1024 * 1024;
export const convexImport = new Command("import").summary("Import data from a file to your deployment").description(
  "Import data from a file to your Convex deployment.\n\n  From a snapshot: `npx convex import snapshot.zip`\n  For a single table: `npx convex import --table tableName file.json`\n\nBy default, this imports into your dev deployment."
).addOption(
  new Option(
    "--table <table>",
    "Destination table name. Required if format is csv, jsonLines, or jsonArray. Not supported if format is zip."
  )
).addOption(
  new Option(
    "--replace",
    "Replace all existing data in any of the imported tables"
  ).conflicts("--append")
).addOption(
  new Option(
    "--append",
    "Append imported data to any existing tables"
  ).conflicts("--replace")
).option(
  "-y, --yes",
  "Skip confirmation prompt when import leads to deleting existing documents"
).addOption(
  new Option(
    "--format <format>",
    "Input file format. This flag is only required if the filename is missing an extension.\n- CSV files must have a header, and each row's entries are interpreted either as a (floating point) number or a string.\n- JSON files must be an array of JSON objects.\n- JSONLines files must have a JSON object per line.\n- ZIP files must have one directory per table, containing <table>/documents.jsonl. Snapshot exports from the Convex dashboard have this format."
  ).choices(["csv", "jsonLines", "jsonArray", "zip"])
).addOption(
  new Option(
    "--component-path <path>",
    "Path to the component in the component tree defined in convex.config.ts"
    // TODO(ENG-6967): Remove hideHelp before launching components
  ).hideHelp()
).addDeploymentSelectionOptions(actionDescription("Import data into")).argument("<path>", "Path to the input file").showHelpAfterError().action(async (filePath, options, command) => {
  const ctx = oneoffContext();
  if (command.args.length > 1) {
    return await ctx.crash({
      exitCode: 1,
      errorType: "fatal",
      printedMessage: `Error: Too many positional arguments. If you're specifying a table name, use the \`--table\` option.`
    });
  }
  const deploymentSelection = deploymentSelectionFromOptions(options);
  const {
    adminKey,
    url: deploymentUrl,
    deploymentName
  } = await fetchDeploymentCredentialsProvisionProd(ctx, deploymentSelection);
  if (!ctx.fs.exists(filePath)) {
    return await ctx.crash({
      exitCode: 1,
      errorType: "invalid filesystem data",
      printedMessage: `Error: Path ${chalk.bold(filePath)} does not exist.`
    });
  }
  const format = await determineFormat(ctx, filePath, options.format ?? null);
  const tableName = options.table ?? null;
  if (tableName === null) {
    if (format !== "zip") {
      return await ctx.crash({
        exitCode: 1,
        errorType: "fatal",
        printedMessage: `Error: The \`--table\` option is required for format ${format}`
      });
    }
  } else {
    if (format === "zip") {
      return await ctx.crash({
        exitCode: 1,
        errorType: "fatal",
        printedMessage: `Error: The \`--table\` option is not allowed for format ${format}`
      });
    }
  }
  await ensureHasConvexDependency(ctx, "import");
  const convexClient = new ConvexHttpClient(deploymentUrl);
  convexClient.setAdminAuth(adminKey);
  const existingImports = await convexClient.query(
    makeFunctionReference(
      "_system/cli/queryImport:list"
    ),
    {}
  );
  const ongoingImports = existingImports.filter(
    (i) => i.state.state === "in_progress"
  );
  if (ongoingImports.length > 0) {
    await askToConfirmImportWithExistingImports(
      ctx,
      deploymentName,
      options.yes
    );
  }
  const fileStats = ctx.fs.stat(filePath);
  showSpinner(ctx, `Importing ${filePath} (${formatSize(fileStats.size)})`);
  let mode = "requireEmpty";
  if (options.append) {
    mode = "append";
  } else if (options.replace) {
    mode = "replace";
  }
  const importArgs = {
    tableName: tableName === null ? void 0 : tableName,
    componentPath: options.componentPath,
    mode,
    format
  };
  const deploymentNotice = options.prod ? ` in your ${chalk.bold("prod")} deployment` : "";
  const tableNotice = tableName ? ` to table "${chalk.bold(tableName)}"` : "";
  const onFailure = async () => {
    logFailure(
      ctx,
      `Importing data from "${chalk.bold(
        filePath
      )}"${tableNotice}${deploymentNotice} failed`
    );
  };
  const importId = await uploadForImport(ctx, {
    deploymentUrl,
    adminKey,
    filePath,
    importArgs,
    onImportFailed: onFailure
  });
  changeSpinner(ctx, "Parsing uploaded data");
  const onProgress = (ctx2, state, checkpointCount) => {
    stopSpinner(ctx2);
    while ((state.checkpoint_messages?.length ?? 0) > checkpointCount) {
      logFinishedStep(ctx2, state.checkpoint_messages[checkpointCount]);
      checkpointCount += 1;
    }
    showSpinner(ctx2, state.progress_message ?? "Importing");
    return checkpointCount;
  };
  while (true) {
    const snapshotImportState = await waitForStableImportState(ctx, {
      importId,
      deploymentUrl,
      adminKey,
      onProgress
    });
    switch (snapshotImportState.state) {
      case "completed":
        logFinishedStep(
          ctx,
          `Added ${snapshotImportState.num_rows_written} documents${tableNotice}${deploymentNotice}.`
        );
        return;
      case "failed":
        return await ctx.crash({
          exitCode: 1,
          errorType: "fatal",
          printedMessage: `Importing data from "${chalk.bold(
            filePath
          )}"${tableNotice}${deploymentNotice} failed

${chalk.red(snapshotImportState.error_message)}`
        });
      case "waiting_for_confirmation": {
        stopSpinner(ctx);
        await askToConfirmImport(
          ctx,
          snapshotImportState.message_to_confirm,
          snapshotImportState.require_manual_confirmation,
          options.yes
        );
        showSpinner(ctx, `Importing`);
        await confirmImport(ctx, {
          importId,
          adminKey,
          deploymentUrl,
          onError: async () => {
            logFailure(
              ctx,
              `Importing data from "${chalk.bold(
                filePath
              )}"${tableNotice}${deploymentNotice} failed`
            );
          }
        });
        break;
      }
      case "uploaded": {
        return await ctx.crash({
          exitCode: 1,
          errorType: "fatal",
          printedMessage: `Import canceled while parsing uploaded file`
        });
      }
      case "in_progress": {
        return await ctx.crash({
          exitCode: 1,
          errorType: "fatal",
          printedMessage: `WARNING: Import is continuing to run on the server. Visit ${snapshotImportDashboardLink(deploymentName)} to monitor its progress.`
        });
      }
      default: {
        const _ = snapshotImportState;
        return await ctx.crash({
          exitCode: 1,
          errorType: "fatal",
          printedMessage: `unknown error: unexpected state ${snapshotImportState}`,
          errForSentry: `unexpected snapshot import state ${snapshotImportState.state}`
        });
      }
    }
  }
});
async function askToConfirmImport(ctx, messageToConfirm, requireManualConfirmation, yes) {
  if (!messageToConfirm?.length) {
    return;
  }
  logMessage(ctx, messageToConfirm);
  if (requireManualConfirmation !== false && !yes) {
    const confirmed = await promptYesNo(ctx, {
      message: "Perform import?",
      default: true
    });
    if (!confirmed) {
      return await ctx.crash({
        exitCode: 1,
        errorType: "fatal",
        printedMessage: "Import canceled"
      });
    }
  }
}
function snapshotImportDashboardLink(deploymentName) {
  return deploymentName === void 0 ? "https://dashboard.convex.dev/deployment/settings/snapshots" : deploymentDashboardUrlPage(deploymentName, "/settings/snapshots");
}
async function askToConfirmImportWithExistingImports(ctx, deploymentName, yes) {
  logMessage(
    ctx,
    `There is already a snapshot import in progress. You can view its progress at ${snapshotImportDashboardLink(deploymentName)}.`
  );
  if (yes) {
    return;
  }
  const confirmed = await promptYesNo(ctx, {
    message: "Start another import?",
    default: true
  });
  if (!confirmed) {
    return await ctx.crash({
      exitCode: 1,
      errorType: "fatal",
      printedMessage: "Import canceled"
    });
  }
}
export async function waitForStableImportState(ctx, args) {
  const { importId, deploymentUrl, adminKey, onProgress } = args;
  const [donePromise, onDone] = waitUntilCalled();
  let snapshotImportState;
  let checkpointCount = 0;
  await subscribe(
    ctx,
    deploymentUrl,
    adminKey,
    "_system/cli/queryImport",
    { importId },
    void 0,
    donePromise,
    {
      onChange: (value) => {
        snapshotImportState = value.state;
        switch (snapshotImportState.state) {
          case "waiting_for_confirmation":
          case "completed":
          case "failed":
            onDone();
            break;
          case "uploaded":
            return;
          case "in_progress":
            checkpointCount = onProgress(
              ctx,
              snapshotImportState,
              checkpointCount
            );
            return;
        }
      }
    }
  );
  return snapshotImportState;
}
async function determineFormat(ctx, filePath, format) {
  const fileExtension = path.extname(filePath);
  if (fileExtension !== "") {
    const formatToExtension = {
      csv: ".csv",
      jsonLines: ".jsonl",
      jsonArray: ".json",
      zip: ".zip"
    };
    const extensionToFormat = Object.fromEntries(
      Object.entries(formatToExtension).map((a) => a.reverse())
    );
    if (format !== null && fileExtension !== formatToExtension[format]) {
      logWarning(
        ctx,
        chalk.yellow(
          `Warning: Extension of file ${filePath} (${fileExtension}) does not match specified format: ${format} (${formatToExtension[format]}).`
        )
      );
    }
    format ?? (format = extensionToFormat[fileExtension] ?? null);
  }
  if (format === null) {
    return await ctx.crash({
      exitCode: 1,
      errorType: "fatal",
      printedMessage: "No input file format inferred by the filename extension or specified. Specify your input file's format using the `--format` flag."
    });
  }
  return format;
}
export async function confirmImport(ctx, args) {
  const { importId, adminKey, deploymentUrl } = args;
  const fetch = deploymentFetch(deploymentUrl, adminKey);
  const performUrl = `/api/perform_import`;
  try {
    await fetch(performUrl, {
      method: "POST",
      body: JSON.stringify({ importId })
    });
  } catch (e) {
    await args.onError(e);
    return await logAndHandleFetchError(ctx, e);
  }
}
export async function uploadForImport(ctx, args) {
  const { deploymentUrl, adminKey, filePath } = args;
  const fetch = deploymentFetch(deploymentUrl, adminKey);
  const data = ctx.fs.createReadStream(filePath, {
    highWaterMark: CHUNK_SIZE
  });
  const fileStats = ctx.fs.stat(filePath);
  showSpinner(ctx, `Importing ${filePath} (${formatSize(fileStats.size)})`);
  let importId;
  try {
    const startResp = await fetch("/api/import/start_upload", {
      method: "POST"
    });
    const { uploadToken } = await startResp.json();
    const partTokens = [];
    let partNumber = 1;
    for await (const chunk of data) {
      const partUrl = `/api/import/upload_part?uploadToken=${encodeURIComponent(
        uploadToken
      )}&partNumber=${partNumber}`;
      const partResp = await fetch(partUrl, {
        headers: {
          "Content-Type": "application/octet-stream"
        },
        body: chunk,
        method: "POST"
      });
      partTokens.push(await partResp.json());
      partNumber += 1;
      changeSpinner(
        ctx,
        `Uploading ${filePath} (${formatSize(data.bytesRead)}/${formatSize(
          fileStats.size
        )})`
      );
    }
    const finishResp = await fetch("/api/import/finish_upload", {
      body: JSON.stringify({
        import: args.importArgs,
        uploadToken,
        partTokens
      }),
      method: "POST"
    });
    const body = await finishResp.json();
    importId = body.importId;
  } catch (e) {
    await args.onImportFailed(e);
    return await logAndHandleFetchError(ctx, e);
  }
  return importId;
}
